### PR TITLE
use host check for entity script whitelist instead of startsWith

### PR DESCRIPTION
**Pull request by birarda**
- fixed a potential entity script whitelist exploit by using a full host check instead of startsWith

### CODE REVIEW

**Comment by birarda** [comment]
#### Test Plan

test script -> https://gist.githubusercontent.com/birarda/641b466704dac409b53aa75d89935ac3/raw/78a4652caa63b8afad31e022695e35106100e50e/test-ess-1.js

(1) Add "gist.githubusercontent" to your script whitelist in domain-server settings. Add an entity that uses the test script. On master this script should run (since gist.githubusercontent.com starts with gist.githubusercontent) but on this PR it should be rejected, since the host does not match exactly.
(2) Smoke test this PR to ensure that scripts from multiple domains you add to the allowed this are allowed, and that all scripts are allowed when the whitelist is clear.

**Comment by hifi-gustavo** [comment]
Build console output for commit: db398b7850782bd0942ffd58c2a3889d9f3ef8df available at: https://gist.github.com/40df5f1d1faa4f443075c7ec563748ab<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9479-db398b7.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9479-db398b7.dmg

**Comment by birarda** [comment]
If edit logging is not enabled for your entity-server you may need to run a second client to determine when the edits are rejected or accepted.

**Comment by OmegaHeron** [comment]
Working on testing now.  I think I've got good telemetry :)

**Comment by OmegaHeron** [comment]
Testing Windows 10;

With this PR and entity scripts allowed from = gist.githubusercontent

I'm not sure I'm getting this right... it seems the script put in an entity is loading up/running in this PR and master even with not fully formed whitelist string.

I created an entity and set its script to the URL above - Interface logs;

`[01/23 17:08:19] [DEBUG] [hifi.scriptengine] Found script in cache: "https://gist.githubusercontent.com/birarda/641b466704dac409b53aa75d89935ac3/raw/78a4652caa63b8afad31e022695e35106100e50e/test-ess-1.js"`
`[01/23 17:08:19] [DEBUG] [hifi.scriptengine.script] script:print()<< Preload 1:  {5aeb31eb-2bd0-4a1f-9992-a7357bb75038}`
`[01/23 17:08:19] [DEBUG] [hifi.scriptengine.script] script:print()<< Unload 1:  {5aeb31eb-2bd0-4a1f-9992-a7357bb75038}`
`[01/23 17:08:19] [CRITICAL] [hifi.scriptengine] "[UncaughtException] Error: Function.prototype.disconnect: failed to disconnect from ScriptEngine::update(float) in Entities 1:9\n[Backtrace]\n    <anonymous>('{5aeb31eb-2bd0-4a1f-9992-a7357bb75038}', ) at https://gist.githubusercontent.com/birarda/641b466704dac409b53aa75d89935ac3/raw/78a4652caa63b8afad31e022695e35106100e50e/test-ess-1.js:9\n    <global>() at -1"`

Interface (Master) logs;
`[01/23 17:16:19] [DEBUG] [hifi.scriptengine] Found script in cache: "https://gist.githubusercontent.com/birarda/641b466704dac409b53aa75d89935ac3/raw/78a4652caa63b8afad31e022695e35106100e50e/test-ess-1.js"`
`[01/23 17:16:19] [DEBUG] [hifi.scriptengine.script] script:print()<< Preload 1:  {55eef3ef-80cc-405c-a0b3-026e54e7ad52}`
`[01/23 17:16:19] [DEBUG] [hifi.scriptengine.script] script:print()<< Unload 1:  {55eef3ef-80cc-405c-a0b3-026e54e7ad52}`
`[01/23 17:16:19] [CRITICAL] [hifi.scriptengine] "[UncaughtException] Error: Function.prototype.disconnect: failed to disconnect from ScriptEngine::update(float) in Entities 1:9\n[Backtrace]\n    <anonymous>('{55eef3ef-80cc-405c-a0b3-026e54e7ad52}', ) at https://gist.githubusercontent.com/birarda/641b466704dac409b53aa75d89935ac3/raw/78a4652caa63b8afad31e022695e35106100e50e/test-ess-1.js:9\n    <global>() at -1"`

**Comment by OmegaHeron** [comment]
Double checking running DS/AC is really from this PR, just in case.

**Comment by OmegaHeron** [comment]
Confirmed my stack on localhost is 9479 via versions listed in nodelist display, my interface is 9479 and I see script print preload/unload with whitelist set to gist.githubusercontent

Maybe I'm having a - it's getting late on a Monday moment and not running test correct or something.

**Comment by birarda** [comment]
@OmegaHeron I believe that may be true for the client that sets the script to start since it sets that change in its local tree.

If you subsequently get an update for that entity from the server it might clear that script but otherwise the only way to see if the script update is not accepted might be to have another client running.

**Comment by OmegaHeron** [comment]
ok - ty @birarda, now it makes sense.  I'll try it again with second interface agent in about an hour (dinner time) unless there's a rush on and you can draft one of the other guys to check it.

**Comment by birarda** [comment]
@ZappoMan can you clarify if that is correct for the Interface client that sets a non whitelisted script?

File: libraries/entities/src/EntityTree.cpp [hunk 1]
```diff
@@ -958,9 +958,12 @@ int EntityTree::processEditPacketData(ReceivedMessage& message, const unsigned c
...
             if (validEditPacket && !_entityScriptSourceWhitelist.isEmpty() && !properties.getScript().isEmpty()) {
                 bool passedWhiteList = false;
-                auto entityScript = properties.getScript();
+
+                // grab a URL representation of the entity script so we can check the host for this script
+                auto entityScriptURL = QUrl::fromUserInput(properties.getScript());
+
                 for (const auto& whiteListedPrefix : _entityScriptSourceWhitelist) {
-                    if (entityScript.startsWith(whiteListedPrefix, Qt::CaseInsensitive)) {
+                    if (entityScriptURL.host().compare(whiteListedPrefix, Qt::CaseInsensitive) == 0) {
```
Comment (ZappoMan) [inline libraries/entities/src/EntityTree.cpp]: the original intent of the feature was to not just compare host names, but also compare the path to the host... how are you handling that?

**Comment by birarda** [comment]
@OmegaHeron I need to fix something about the way this works and then I'll have you re-test!

**Comment by OmegaHeron** [comment]
ok @birarda - having internet problems here anyhow - wind and tree limbs + cable lines.  If I go dark and don't reply feel free to hand it off to anyone else around to test.  Otherwise will return to it when ready.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 908464e1236c8dcc42a7d8a839eec4bc41550be8 available at: https://gist.github.com/40df5f1d1faa4f443075c7ec563748ab<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9479-908464e.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9479-908464e.dmg

File: libraries/entities/src/EntityTree.cpp [hunk 2]
```diff
@@ -958,9 +958,15 @@ int EntityTree::processEditPacketData(ReceivedMessage& message, const unsigned c
...
             if (validEditPacket && !_entityScriptSourceWhitelist.isEmpty() && !properties.getScript().isEmpty()) {
                 bool passedWhiteList = false;
-                auto entityScript = properties.getScript();
+
+                // grab a URL representation of the entity script so we can check the host for this script
+                auto entityScriptURL = QUrl::fromUserInput(properties.getScript());
+
                 for (const auto& whiteListedPrefix : _entityScriptSourceWhitelist) {
-                    if (entityScript.startsWith(whiteListedPrefix, Qt::CaseInsensitive)) {
+                    auto whiteListURL = QUrl::fromUserInput(_entityScriptSourceWhitelist);
```
Comment (ZappoMan) [inline libraries/entities/src/EntityTree.cpp]: don't you mean `whiteListedPrefix` here?

**Comment by birarda** [comment]
@OmegaHeron I have updated the test plan for this PR. I will check with qatesters to see if someone else can jump on it otherwise if you can grab it when you get back that would be great. Be safe!

**Comment by nbq** [comment]
Will test once it builds

**Comment by hifi-gustavo** [comment]
Build console output for commit: 9229a07f6a3f0ec34647535d3e1a0a93219272c0 available at: https://gist.github.com/40df5f1d1faa4f443075c7ec563748ab<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9479-9229a07.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9479-9229a07.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 7362bf16c1d2b1be4e1ec9c5d56523f68afbad72 available at: https://gist.github.com/40df5f1d1faa4f443075c7ec563748ab<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9479-7362bf1.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9479-7362bf1.dmg

**Comment by birarda** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: 9229a07f6a3f0ec34647535d3e1a0a93219272c0 available at: https://gist.github.com/40df5f1d1faa4f443075c7ec563748ab<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9479-9229a07.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9479-9229a07.dmg

**Comment by nbq** [comment]
**PR FAIL**
---
Tester: Coal
CPU: i7 6700k
GPU: GF 970
OS: Windows 10

1. Added `gist.githubusercontent` to the whitelist under `Entity Scripts Allowed from` - Ran script provided for this PR and checked and it did not run as there was no mention of `Preload` or `Unload` in the logs.
2. Changed whitelist to include `gist.githubusercontent.com` and ran the script in an entity and noticed this in the logs so I know it ran: `[01/23 21:01:25] [DEBUG] [hifi.scriptengine.script] script:print()<< Preload 1:  {ef6237bb-ff63-4ab8-847f-f93b30dac6d0}` - It also triggered this on deletion: `[01/23 21:02:18] [DEBUG] [hifi.scriptengine.script] script:print()<< Unload 1:  {ef6237bb-ff63-4ab8-847f-f93b30dac6d0}`
3. Changed whitelist to be `gist.githubusercontent.com/birarda` and I ran the script in an entity and received the same Preload and Unload as before.
4. Next I changed the whitelist to be `gist.githubusercontent.com/Atlante45` and ran the script and it DID run the script, though it said it was cached and ran the Preload and Unload immediately. Shown below in the logs (1) - THIS IS A FAIL


(1)
```
[01/23 21:06:42] [DEBUG] [hifi.scriptengine] Found script in cache: "https://gist.githubusercontent.com/birarda/641b466704dac409b53aa75d89935ac3/raw/61c75befce101f66b84b87e244b51f4b001dc3e5/test-ess-1.js"
[01/23 21:06:42] [DEBUG] [hifi.scriptengine.script] script:print()<< Preload 1:  {a030aa95-e585-4117-bc3d-b2f2cd100a4b}
[01/23 21:06:42] [DEBUG] [hifi.scriptengine.script] script:print()<< Unload 1:  {a030aa95-e585-4117-bc3d-b2f2cd100a4b}
```

**Comment by birarda** [comment]
@nbq I think it will be difficult to test with only a single client. It is possible that the single client runs the script temporarily before getting a new edit from the entity server that tells it there is no longer a script to run.

If you enable edit logging in your domain server settings the entity server will log out when it rejects an edit because of the script whitelist. Alternatively if you can setup a second client you will be able to see if it gets the script changes.

**Comment by nbq** [comment]
Ok, I will retry this, thanks @birarda

**Comment by nbq** [comment]
## Test 2 - FAIL STILL
Tester: Coal
CPU: i7 6700k
GPU: GF 970
OS: Windows 10

1. Added `gist.githubusercontent` to the whitelist under `Entity Scripts Allowed from` - Ran script provided for this PR and checked and it did not run as there was no mention of `Preload` or `Unload` in the logs.
2. Changed whitelist to include `gist.githubusercontent.com` and ran the script in an entity and noticed this in the logs so I know it ran: `[01/23 21:01:25] [DEBUG] [hifi.scriptengine.script] script:print()<< Preload 1:  {ef6237bb-ff63-4ab8-847f-f93b30dac6d0}` - It also triggered this on deletion: `[01/23 21:02:18] [DEBUG] [hifi.scriptengine.script] script:print()<< Unload 1:  {ef6237bb-ff63-4ab8-847f-f93b30dac6d0}`
3. Changed whitelist to be `gist.githubusercontent.com/birarda` and I ran the script in an entity and received the same Preload and Unload as before.
4. Next I changed the whitelist to be `gist.githubusercontent.com/Atlante45` - **Edited Part** With two Interfaces open, the new Interface did not load the script while the old interface did. This is even with `Edit Logging` turned on. *This means users who have been on the domain before where there was a white list and then edited can still run scripts from all locations*
5. Tested a script from a domain not on the list and it did not run on either client. Added that domain to the whitelist and it ran perfectly fine on both Interfaces. 
6. I cleaned the domain whitelist so it was empty attempted to run a script from another web site and only the Interface with the original `whitelist empty` setting ran that script.

**QA FAIL still as you can see on test 5 and 6, it seems the Interfaces are never getting/following updated whitelists, just the ones that exist when the Interface first pulls the settings.**

**Comment by birarda** [comment]
@nbq the script whitelist is only respected by the entity server. What you should expect to see is that the entity server refuses to accept edits that include scripts that are not on the whitelist.

When testing with two interface clients you should see that the client that is NOT changing the script URL does not receive changes to the URL for a script that does not match the whitelist.

**Comment by nbq** [comment]
## Retest notes 2

After going through the tests I see what you mean @birarda because when something is on the white list, it goes through on both clients. When something is not on the white list it just runs local and is never passed onto the other Interface.

With that now being tested I updated this to a PASS after testing with domains not on the white list and domains on the white list. Only the domains on the white list get passed through to the other Interface.

**Pull request merged**